Sources/UtilityBeltData/CoreData/CoreDataManager.swift

```diff
@@ -0,0 +1,157 @@
+        fetchRequest.includesPropertyValues = false
+        fetchRequest.includesSubentities = false
+        fetchRequest.predicate = predicate
+
+        let count = try self.managedContext().count(for: fetchRequest)
+
+        // If the fetched objects were not found, ensure that we return 0
+        return (count != NSNotFound) ? count : 0
+    }
+
+    // MARK: Create
+
+    /// Creates a new instance of a managed object subclass.
+    /// - Parameter type: The managed object subclass type to create.
+    /// - Parameter context: The managed object context to create the object in. If nil, uses the current default context.
+    public func newInstance<T: NSManagedObject>(of type: T.Type, in context: NSManagedObjectContext? = nil) throws -> T? {
+        let context = try (context ?? self.managedContext())
+
+        return NSEntityDescription.insertNewObject(forEntityName: String(describing: T.self), into: context) as? T
+    }
+
+    // MARK: Delete
+
+    /// Deletes a single object.
+    /// - Parameter object: The object to delete.
+    public func delete<T: NSManagedObject>(_ object: T) throws {
+        let context = try (object.managedObjectContext ?? self.managedContext())
+
+        context.delete(object)
+        try context.save()
+    }
+
+    /// Deletes all objects of a given entity type.
+    /// - Parameter type: The entity type to batch delete.
+    ///
+    /// Batch delete requests are only compatible with the SQLite store type.
+    public func deleteAll<T: NSManagedObject>(of type: T.Type,
+                                              with predicate: NSPredicate? = nil) throws {
+        let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: String(describing: T.self))
+        fetchRequest.predicate = predicate
+
+        let deleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)
+
+        try self.managedContext().executeAndMergeChanges(using: deleteRequest)
+
+        try self.managedContext().save()
+    }
+
+    // MARK: Exists
+
+    /// Checks whether a given type of entity exists.
+    /// - Parameter type: The type of entity to check existence for.
+    public func exists<T: NSManagedObject>(_ type: T.Type,
+                                           with predicate: NSPredicate? = nil) throws -> Bool {
+        return try self.count(of: type, with: predicate) > 0
+    }
+
+    // MARK: Fetch
+
+    /// Fetches all managed objects of a given type.
+    /// - Parameter type: The type of entity to fetch all results for.
+    public func fetchAll<T: NSManagedObject>(of type: T.Type,
+                                             with predicate: NSPredicate? = nil,
+                                             in context: NSManagedObjectContext? = nil) throws -> [T] {
+        let context = try (context ?? self.managedContext())
+
+        // Instead of using T.fetchRequest(), we build the FetchRequest so we don't need to cast the result
```

Comment:
will T.fetchRequest() even work here? seems like it wasnt working from your comment before?

Reply:
Updated the other `TODO`, both should reflect the same decision here for anyone new to CoreData to understand why `T.fetchRequest()` isn't being used.